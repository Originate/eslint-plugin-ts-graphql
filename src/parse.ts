import { TSESLint, TSESTree } from "@typescript-eslint/experimental-utils";
import { MessageId } from ".";

export interface QuickDocument {
  operationName: string | null;
  operationType: string;
}

const operationNameRegex = /^\s*(query|mutation|subscription|fragment)[\s\r\n]*([_A-Za-z][_0-9A-Za-z]*)?/gim;

// This is not a real parser - just a quick-and-dirty regex.
function parseOperationNames(node: TSESTree.TemplateLiteral): QuickDocument[] {
  return node.quasis.flatMap((quasi) => {
    const matches = operationNameRegex[Symbol.matchAll](quasi.value.cooked);
    return Array.from(matches).map(([, operationType, operationName]) => ({
      operationName,
      operationType,
    }));
  });
}

export function handleTemplateTag(
  node: TSESTree.TaggedTemplateExpression,
  report: TSESLint.RuleContext<MessageId, any>["report"]
): QuickDocument | undefined {
  const documents = parseOperationNames(node.quasi);
  if (documents.length > 1) {
    report({
      messageId: "singleOperation",
      node,
    });
    return;
  }
  if (documents.length === 0) {
    report({
      messageId: "operationOrSingleFragment",
      node,
    });
    return;
  }
  return documents[0];
}
