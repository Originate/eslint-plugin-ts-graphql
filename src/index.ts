import {
  ESLintUtils,
  TSESLint,
  TSESTree,
} from "@typescript-eslint/experimental-utils";
import { isMatch } from "lodash";
import { getNodes } from "./get-nodes";
import { handleTemplateTag, QuickDocument } from "./parse";
import { dasherize } from "./utils";

// TODO:
const createRule = ESLintUtils.RuleCreator(
  (name) =>
    `https://github.com/Thinkmill/ts-gql/blob/master/packages/eslint-plugin/docs/rules/${name}.md`
);

const messages = {
  singleOperation: "GraphQL documents must have only one operation",
  mustBeNamed: "GraphQL operations must have a name",
  mustUseAs: "You must cast gql tags with the generated type",
  mustUseAsCorrectly:
    "The type assertion on a gql tag is not in the expected format",
  operationOrSingleFragment:
    "GraphQL documents must either have a single operation or a single fragment",
};

export type MessageId = keyof typeof messages;

function checkFragment(
  document: QuickDocument,
  node: TSESTree.TaggedTemplateExpression,
  context: TSESLint.RuleContext<MessageId, any>
) {
  addNameToGqlTag(document, node, context);
}

function addNameToGqlTag(
  document: QuickDocument,
  node: TSESTree.TaggedTemplateExpression,
  context: TSESLint.RuleContext<MessageId, any>
) {
  if (!document.operationName) {
    context.report({
      messageId: "mustBeNamed",
      node,
    });
    return false;
  }

  if (document.operationType === "fragment") {
    // Fragments do not need a type assertion
    return true;
  }

  const name = document.operationName;
  const pathname = `./__generated__/${dasherize(name)}`;
  const sanitizedPathname = JSON.stringify(pathname);

  const typedDocumentNodePackageName = "@graphql-typed-document-node/core";
  const type = `import(${JSON.stringify(
    typedDocumentNodePackageName
  )}).TypedDocumentNode<
  import(${sanitizedPathname}).${name},
  import(${sanitizedPathname}).${name}Variables
>`;

  // Add a type assertion if there isn't one.
  if (node.parent?.type !== "TSAsExpression") {
    context.report({
      messageId: "mustUseAs",
      node,
      fix(fix) {
        return fix.insertTextAfter(node, ` as ${type}`);
      },
    });
    return false;
  }

  // The gql template has a type assertion - check if it has the correct format.
  if (
    !isMatch(node.parent.typeAnnotation, {
      type: "TSImportType",
      parameter: {
        literal: {
          value: typedDocumentNodePackageName,
        },
      },
      qualifier: {
        type: "Identifier",
        name: "TypedDocumentNode",
      },
      typeParameters: {
        params: [
          {
            type: "TSImportType",
            parameter: {
              literal: {
                value: pathname,
              },
            },
            qualifier: {
              type: "Identifier",
              name: name,
            },
          },
          {
            type: "TSImportType",
            parameter: {
              literal: {
                value: pathname,
              },
            },
            qualifier: {
              type: "Identifier",
              name: `${name}Variables`,
            },
          },
        ],
      },
    })
  ) {
    const typeAnnotation = node.parent.typeAnnotation;
    context.report({
      messageId: "mustUseAsCorrectly",
      node: typeAnnotation,
      fix(fix) {
        return fix.replaceText(typeAnnotation, type);
      },
    });
    return false;
  }
  return true;
}

function checkDocument(
  document: QuickDocument,
  node: TSESTree.TaggedTemplateExpression,
  context: TSESLint.RuleContext<MessageId, any>
) {
  if (document.operationType === "fragment") {
    checkFragment(document, node, context);
    return;
  }
  addNameToGqlTag(document, node, context);
}

export const rules = {
  "gql-type-assertion": createRule<[], MessageId>({
    name: "gql-type-assertion",
    meta: {
      fixable: "code",
      docs: {
        requiresTypeChecking: true,
        category: "Best Practices",
        recommended: "error",
        description:
          "Fixing this error adds a type assertion to gql template tags that provides accurate type inference for query variables, and for result data.",
      },
      messages,
      type: "problem",
      schema: [],
    },
    defaultOptions: [],
    create(context) {
      return {
        Program(programNode) {
          let report: typeof context["report"] = (arg) => {
            return context.report(arg);
          };

          for (const node of getNodes(context, programNode)) {
            if (node.type === "TaggedTemplateExpression") {
              if (node.tag.type === "Identifier" && node.tag.name === "gql") {
                const document = handleTemplateTag(node, report);
                if (document) {
                  checkDocument(document, node, context);
                }
              }
            }
          }
        },
      };
    },
  }),
};
