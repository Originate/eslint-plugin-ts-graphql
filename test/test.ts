import { SnapshotCreator } from "eslint-snapshot-test";
import fs from "fs";
import path from "path";
import { rules } from "../src";

let snapshotCreator = new SnapshotCreator({
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2020,
    ecmaFeatures: {
      jsx: true,
    },
    sourceType: "module",
  },
});

let fixturesPath = path.join(__dirname, "__fixtures__");

let fixtures = fs.readdirSync(fixturesPath);

for (const fixture of fixtures) {
  test(fixture, async () => {
    const fixturePath = path.join(fixturesPath, fixture);
    const code = fs.readFileSync(fixturePath, "utf8");
    let result = snapshotCreator
      .mark({
        code,
        rule: rules["gql-type-assertion"],
        ruleName: "gql-type-assertion",
      })
      .withFileName(fixturePath)
      .render();
    if (result.fixedOutput === code) {
      delete result.fixedOutput;
    }
    expect(result).toMatchSnapshot();
  });
}
