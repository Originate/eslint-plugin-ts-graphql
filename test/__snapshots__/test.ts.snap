// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"gql\\";

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>;
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "fix": Object {
        "range": Array [
          68,
          68,
        ],
        "text": " as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>",
      },
      "line": 3,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"gql\\";

gql\`
~~~~    [You must cast gql tags with the generated type]
  query Thing {
~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    something
~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`;
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`basic-fixed.ts 1`] = `
Object {
  "lintMessages": Array [],
  "snapshot": "
import { gql } from \\"gql\\";

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>;
",
}
`;

exports[`fragment.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"gql\\";

const fields = gql\`
  fragment Fields on Thing {
    id
    something
  }
\`

gql\`
  \${fields}
  query GetThing {
    ...Fields
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/get-thing\\").GetThing,
  import(\\"./__generated__/get-thing\\").GetThingVariables
>
",
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 15,
      "fix": Object {
        "range": Array [
          160,
          160,
        ],
        "text": " as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/get-thing\\").GetThing,
  import(\\"./__generated__/get-thing\\").GetThingVariables
>",
      },
      "line": 10,
      "message": "You must cast gql tags with the generated type",
      "messageId": "mustUseAs",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"gql\\";

const fields = gql\`
  fragment Fields on Thing {
    id
    something
  }
\`

gql\`
~~~~    [You must cast gql tags with the generated type]
  \${fields}
~~~~~~~~~~~    [You must cast gql tags with the generated type]
  query GetThing {
~~~~~~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
    ...Fields
~~~~~~~~~~~~~    [You must cast gql tags with the generated type]
  }
~~~    [You must cast gql tags with the generated type]
\`
~    [You must cast gql tags with the generated type]
",
}
`;

exports[`invalid-type-assertion.ts 1`] = `
Object {
  "fixedOutput": "import { gql } from \\"gql\\";

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>;

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>;

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>;
",
  "lintMessages": Array [
    Object {
      "column": 6,
      "endColumn": 2,
      "endLine": 10,
      "fix": Object {
        "range": Array [
          72,
          173,
        ],
        "text": "import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>",
      },
      "line": 7,
      "message": "The type assertion on a gql tag is not in the expected format",
      "messageId": "mustUseAsCorrectly",
      "nodeType": "TSTypeReference",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
    Object {
      "column": 6,
      "endColumn": 2,
      "endLine": 19,
      "fix": Object {
        "range": Array [
          220,
          379,
        ],
        "text": "import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>",
      },
      "line": 16,
      "message": "The type assertion on a gql tag is not in the expected format",
      "messageId": "mustUseAsCorrectly",
      "nodeType": "TSImportType",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
    Object {
      "column": 6,
      "endColumn": 2,
      "endLine": 28,
      "fix": Object {
        "range": Array [
          426,
          586,
        ],
        "text": "import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
  import(\\"./__generated__/thing\\").Thing,
  import(\\"./__generated__/thing\\").ThingVariables
>",
      },
      "line": 25,
      "message": "The type assertion on a gql tag is not in the expected format",
      "messageId": "mustUseAsCorrectly",
      "nodeType": "TSImportType",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"gql\\";

gql\`
  query Thing {
    something
  }
\` as SomeType<
     ~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/thing\\").Thing,
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/thing\\").ThingVariables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
>;
~    [The type assertion on a gql tag is not in the expected format]

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/thing\\").WrongThing,
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/thing\\").ThingVariables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
>;
~    [The type assertion on a gql tag is not in the expected format]

gql\`
  query Thing {
    something
  }
\` as import(\\"@graphql-typed-document-node/core\\").TypedDocumentNode<
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/thing\\").Thing,
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
  import(\\"./__generated__/wrong-thing\\").ThingVariables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    [The type assertion on a gql tag is not in the expected format]
>;
~    [The type assertion on a gql tag is not in the expected format]
",
}
`;

exports[`multiple-operations.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 10,
      "line": 3,
      "message": "GraphQL documents must have only one operation",
      "messageId": "singleOperation",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"gql\\";

gql\`
~~~~    [GraphQL documents must have only one operation]
  query{
~~~~~~~~    [GraphQL documents must have only one operation]
    something
~~~~~~~~~~~~~    [GraphQL documents must have only one operation]
  }
~~~    [GraphQL documents must have only one operation]
  query Thing {
~~~~~~~~~~~~~~~    [GraphQL documents must have only one operation]
    something
~~~~~~~~~~~~~    [GraphQL documents must have only one operation]
  }
~~~    [GraphQL documents must have only one operation]
\`;
~    [GraphQL documents must have only one operation]
",
}
`;

exports[`no-name.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "line": 3,
      "message": "GraphQL operations must have a name",
      "messageId": "mustBeNamed",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [GraphQL operations must have a name]
  query {
~~~~~~~~~    [GraphQL operations must have a name]
    something
~~~~~~~~~~~~~    [GraphQL operations must have a name]
  }
~~~    [GraphQL operations must have a name]
\`;
~    [GraphQL operations must have a name]
",
}
`;

exports[`parse-error.ts 1`] = `
Object {
  "lintMessages": Array [
    Object {
      "column": 1,
      "endColumn": 2,
      "endLine": 7,
      "line": 3,
      "message": "GraphQL documents must either have a single operation or a single fragment",
      "messageId": "operationOrSingleFragment",
      "nodeType": "TaggedTemplateExpression",
      "ruleId": "gql-type-assertion",
      "severity": 2,
    },
  ],
  "snapshot": "
import { gql } from \\"@ts-gql/tag\\";

gql\`
~~~~    [GraphQL documents must either have a single operation or a single fragment]
  que Thing {
~~~~~~~~~~~~~    [GraphQL documents must either have a single operation or a single fragment]
    something
~~~~~~~~~~~~~    [GraphQL documents must either have a single operation or a single fragment]
  }
~~~    [GraphQL documents must either have a single operation or a single fragment]
\`;
~    [GraphQL documents must either have a single operation or a single fragment]
",
}
`;
